heclib/hecdss/hecdss.c

```diff
@@ -0,0 +1,742 @@
+/// <param name="fullPath"></param>
+/// <param name="x"></param>
+/// <param name="y"></param>
+/// <param name="z"></param>
+/// <param name="coordinateSystem"></param>
+/// <param name="coordinateID"></param>
+/// <param name="horizontalUnits"></param>
+/// <param name="horizontalDatum"></param>
+/// <param name="verticalUnits"></param>
+/// <param name="verticalDatum"></param>
+/// <param name="timeZoneName"></param>
+/// <param name="supplemental"></param>
+/// <param name="replace">set to 1 to overwrite existing data, otherwise use 0</param>
+/// <returns></returns>
+HECDSS_API int hec_dss_locationStore(dss_file* dss, const char* fullPath,
+  const double x, const double y, const double z,
+  const int coordinateSystem, const int coordinateID,
+  const int horizontalUnits, const int horizontalDatum,
+  const int verticalUnits, const int verticalDatum,
+  const char* timeZoneName,
+  const char* supplemental,
+  const int replace) {
+
+  zStructLocation* loc = zstructLocationNew(fullPath);
+  
+  loc->xOrdinate = x;
+  loc->yOrdinate = y;
+  loc->zOrdinate = z;
+  loc->coordinateSystem = coordinateSystem;
+  loc->coordinateID = coordinateID;
+  loc->verticalUnits = verticalUnits;
+  loc->verticalDatum = verticalDatum;
+  loc->horizontalDatum = horizontalDatum;
+  loc->horizontalUnits = horizontalUnits;
+  loc->timeZoneName = mallocAndCopy(timeZoneName);
+  loc->allocated[zSTRUCT_timeZoneName] = 1;
+  loc->supplemental = mallocAndCopy(supplemental);
+  loc->allocated[zSTRUCT_otherInformation] = 1;
+  int status = zlocationStore(dss->ifltab, loc, replace);
+
+  zstructFree(loc);
+  return status;
+}
+
+
+/// <summary>
+/// hec_dss_pdRetrieveInfo is used by client app to determine
+/// required array sizes for calling hec_dss_pdRetrieve.
+/// 
+/// </summary>
+/// <param name="dss">pointer to dss_file</param>
+/// <param name="pathname">path of paired data</param>
+/// <param name="numberOrdinates">number of rows for ordinates, and curves</param>
+/// <param name="numberCurves">number of columns in the curve dataset</param>
+/// <returns></returns>
+HECDSS_API int hec_dss_pdRetrieveInfo(dss_file* dss, const char* pathname,
+                                    int* numberOrdinates, int* numberCurves, 
+                                    char* unitsIndependent, const int unitsIndependentLength,
+                                    char* unitsDependent, const int unitsDependentLength,
+                                    char* typeIndependent, const int typeIndependentLength,
+                                    char* typeDependent, const int typeDependentLength,
+                                    int* labelsLength){
+  zStructPairedData* pds = zstructPdNew(pathname);
+  //pds->endingOrdinate = 1;// hack to retrive meta data, with minimal actual data.
+  int returnDoubles = 2;
+  int status = zpdRetrieve(dss->ifltab, pds, returnDoubles);
+
+  *numberOrdinates = pds->numberOrdinates;
+  *numberCurves = pds->numberCurves;
```

Comment:
Initialize char* out parameters below to empty strings either before `stringCopy(...)` or in `else` clauses.